data/tutorials/language/0lg_09_env.md

```diff
@@ -0,0 +1,562 @@
+
+In practice, curried functions are the default form functions should take because:
+- Allows partial application
+- Less editing, no parenthesis and comas
+- No pattern matching over a tuple
+
+### Functions With Side-Effects
+
+With respect to its type, a function is expected to process input data from its domain and produce a result data from its codomain.
+
+However, some functions either take input data outside of their declared domain or produce data outside of their codomain. These out-of-signature data are called effects, or side effects. Input and output are the most common forms of effects. Input is out-of-domain data and output out-of-codomain data. However, the result of functions returning random numbers or the current time is influenced by external factors, which is also called an effect. The external factor is out-of-domain input. Similarly, any observable phenomena triggered by the computation of a function is out-of-codomain output.
+
+In practice, what is considered an effect is an engineering choice. In most circumstances, I/O operations are considered as effects, unless they are ignored. Electromagnetic radiation emitted by the processor when computing a function isn't usually considered a relevant side-effect, except in some security-sensitive contexts. In the OCaml community, as well as in the wider functional programming community, functions are often said to be either pure or impure. The former does not have side effects, the latter does. This distinction makes sense and is useful. Knowing what the effects are, and when are they taking place, is a key design consideration. However, it is important to remember this distinction always assumes some sort of context. Any computation has effects, and what is considered a relevant effect is a design choice.
+
+Since, by definition, effects lie outside function types, a function type can't reflect the effects a function may have. However, it is important to document the intended side effects a function may have. Consider the `Unix.time` function. It returns the number of seconds elapsed since Jan 1, 1970.
+```ocaml
+# Unix.time ;;
+- : unit -> float = <fun>
+```
+(If you're getting an `Unbound module error` in macOS, run this first: `#require "unix";;`.) 
+
+No data needs to be passed to this function in order to return its result. It is entirely context-dependent, i.e., entirely determined by external factors. Therefore, if it was passed something, it would be discarded. Note that it has to be passed something to trigger the computation. Requesting the information from the operating system, as suggested by the module name, is that computation. To sum up, since the function must receive data to trigger the computation but the data is going to be ignored, it makes sense to provide `()`, the unit value, the inhabitant of the type that only has a single value. What gets discarded is meaningless in the first place.
```

Comment:
```suggestion
No data needs to be passed to this function in order to return its result. It is entirely context-dependent, i.e., entirely determined by external factors. Therefore, if it was passed something, it would be discarded. Note that it has to be passed something to trigger the computation. Requesting the information from the operating system, as suggested by the module name, is that computation. To sum up, since the function must receive data to trigger the computation, but the data is going to be ignored, it makes sense to provide `()`, the unit value and the inhabitant of a single-value type. What gets discarded is meaningless in the first place.
```
Tighter sentence. Is this still accurate?